complex/ft_msg.c

```diff
@@ -217,13 +233,17 @@ int ft_send_msg(void)
 {
 	int ret;
 
-	while (!ft_tx.credits) {
-		ret = ft_comp_tx(FT_COMP_TO);
-		if (ret)
-			return ret;
+	if (test_info.class_function != FT_FUNC_INJECT &&
+		test_info.class_function != FT_FUNC_INJECTDATA) {
+		while (!ft_tx.credits) {
+			ret = ft_comp_tx(FT_COMP_TO);
+			if (ret)
+				return ret;
+		}
+
+		ft_tx.credits--;
 	}
 
-	ft_tx.credits--;
```

Comment:
Sounds good. I guess ft_comp_tx(..) has to be wrapped in a condition that checks if the provider is in manual progress mode. Otherwise the call could hang.